gui/tiletypes.lua

```diff
@@ -0,0 +1,1667 @@
+        and dfhack.maps.isValidTilePos(self.last_point.x or -1, self.last_point.y or -1, self.last_point.z or -1)
+    then
+        self.box = Box{
+            self.first_point,
+            self.last_point
+        }
+        if type(self.on_confirm) == "function" then
+            self.on_confirm(self.box)
+        end
+    end
+end
+
+function BoxSelection:clear()
+    self.box = nil
+    self.first_point = nil
+    self.last_point = nil
+end
+
+function BoxSelection:onInput(keys)
+    if BoxSelection.super.onInput(self, keys) then
+        return true
+    end
+    if keys.LEAVESCREEN or keys._MOUSE_R then
+        return false
+    end
+
+    local mousePos = dfhack.gui.getMousePos(true)
+    local cursorPos = copyall(df.global.cursor)
+
+    if keys.SELECT then
+        if self.first_point and not self.last_point then
+            if not self.flat or cursorPos.z == self.first_point.z then
+                df.global.cursor.x = math.max(math.min(cursorPos.x, df.global.world.map.x_count - 1), 0)
+                df.global.cursor.y = math.max(math.min(cursorPos.y, df.global.world.map.y_count - 1), 0)
+                cursorPos = guidm.getCursorPos()
+                self.last_point = cursorPos
+                self:confirm()
+            end
+        elseif dfhack.maps.isValidTilePos(cursorPos.x, cursorPos.y, cursorPos.z) then
+            self.first_point = self.first_point or cursorPos
+        end
+
+        return true
+    end
+
+    local avoid_rect = type(self.avoid_rect) == "function" and self.avoid_rect() or self.avoid_rect
+
+    -- Get the position of the mouse in coordinates local to avoid_rect, if it's specified
+    local mouseFramePos = avoid_rect and self:getMousePos(gui.ViewRect{
+        rect=gui.mkdims_wh(
+            avoid_rect.x1,
+            avoid_rect.y1,
+            avoid_rect.width,
+            avoid_rect.height
+        )
+    })
+
+    if keys._MOUSE_L and not mouseFramePos then
+        -- If left click and the mouse is not in the avoid_rect
+        if self.first_point and not self.last_point then
+            if not self.flat or mousePos.z == self.first_point.z then
+                local inBoundsMouse = {
+                    x = math.max(math.min(mousePos.x, df.global.world.map.x_count - 1), 0),
+                    y = math.max(math.min(mousePos.y, df.global.world.map.y_count - 1), 0),
+                    z = mousePos.z,
+                }
+                self.last_point = inBoundsMouse
+                self:confirm()
+            end
+        elseif dfhack.maps.isValidTilePos(mousePos.x, mousePos.y, mousePos.z) then
+            self.first_point = self.first_point or mousePos
+        end
+
+        return true
+    end
+
+    -- Switch to the cursor if the cursor was moved (Excluding up and down a Z level)
+    local filteredKeys = utils.clone(keys)
+    filteredKeys["CURSOR_DOWN_Z"] = nil
+    filteredKeys["CURSOR_UP_Z"] = nil
+    self.useCursor = self.useCursor or guidm.getMapKey(filteredKeys)
```

Comment:
Could be, I tried to keep them independent to prevent cases where keyboard cursor users move their mouse for some reason and lose their cursor position, but I'm not a keyboard cursor user so I can't really speak for them.

Reply:
I have some similar logic in the DFHack logo context menu. The menu selection can follow the mouse or be controlled by the keyboard. in that one, selection control jumps to the mouse when it moves outside of the item its currently hovering over.

https://github.com/DFHack/dfhack/blob/develop/plugins/lua/hotkeys.lua#L347

maybe here we can be a little less sensitive by detecting when it moves to a different screen interface tile instead of the current code that detects when it moves to a different pixel.